### PR TITLE
Added the advertisement service as a real service

### PR DESCRIPTION
**Pull request by Noodlez1232**
The advertisement service was previously only advertised. This makes it a real service after a discussion in Matrix. Although it's advertised, the service disappears upon connection which might cause confusion.

### CODE REVIEW

**Comment by FlorentRevest** [comment]
As of now, I don't really see the need for this PR. It seems to me that advertising UUIDs and Service UUIDs are different things, why should there be a dummy service with the same UUID as the advertising packet ?

Your description mentions it could "lead to confusion", is this the only reason ? Who is it confusing for ? Some bluetooth software stack or you ? Because I actually find the opposite, that introducing a dummy service with the UUID of the advertising data conflates two different things and makes it more confusing.

If you could elaborate a bit more on the reasoning maybe I'd realize I'm missing something but for now it only seems to add code that isn't useful.

Also if there's a canonical source that recommends having a matching service, it seems that a full fix would be to also advertise the other service uuids ?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ set(SRC
 	descriptor.cpp
 	application.cpp
 	advertisement.cpp
+    advertisementservice.cpp
```
Comment (FlorentRevest) [inline src/CMakeLists.txt:14]: The rest of the file uses tabs

File: src/advertisementservice.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#include <QDBusInterface>
+#include <QDBusMessage>
+
+#include "advertisementservice.h"
+#include "characteristic.h"
+#include "common.h"
+
+AdvertisementService::AdvertisementService(int index, QDBusConnection bus, QObject *parent) : Service(bus, index, ADVERT_UUID, parent)
+{
+    // A stub service
```
Comment (FlorentRevest) [inline src/advertisementservice.cpp:27]: Given that this file solely exists to define an empty constructor and will never evolve, you might as well inline the constructor definition within the class declaration and remove this cpp file

**Review by FlorentRevest — COMMENTED** [review]
I don't understand the need for this patch? You say it can "cause confusion", but how so ?

**Comment by Noodlez1232** [comment]
Apologies for the late reply, got caught up in school.

> As of now, I don't really see the need for this PR. It seems to me that advertising UUIDs and Service UUIDs are different things, why should there be a dummy service with the same UUID as the advertising packet ?

My mindset here was it didn't make sense to advertise a service that didn't actually exist. It seemed a bit nonsensical to me. Which would explain the other two questions as well.

There's another reason, and that's that there are other service disccovery methods that this doesn't take into account (Purely advertising the service). This also includes SDP, where if the service doesn't actually exist, the discovery doesn't happen. This is more of a software problem than a hardware problem, and doesn't *seem* to be against spec, but may disallow other software from discovering it (Software that is unable to do discovery filtering based on advertised services, and only on discovered services).

> Also if there's a canonical source that recommends having a matching service, it seems that a full fix would be to also advertise the other service uuids ?

In my free time in school, I got a bit curious about this and scoured the BT(LE) specs, and it doesn't mention if either way is needed. In fact, it seems that service advertisement is just an advertisement that happens to have a Service UUID in it, and no other information is provided alongside it (That I can find. I went through GATT, GAP, and a bit of LE specs).

So I suppose, since it's not mentioned in the spec, this pull request could be ignored, but keep in mind that services that are purely advertised don't show up during actual connection and service discovery. Only actual provided services show up there.

**Comment by Noodlez1232** [comment]
So to further clarify in a much shorter TL;DR of the rationale here:
Service advertisement at the LE level is just that. An advertisement that happens to contain a service ID. This is fine for filtering when you can filter based on advertisements. (Meaning services that exist don't have to be advertised).
But advertised services *dont* show up during service discovery, which could be a problem for software that cannot filter based on advertised services. (Meaning that advertised services should actually exist, if that's an important thing).

**Comment by FlorentRevest** [comment]
> Apologies for the late reply, got caught up in school.

My turn to apologize! :D 

> > As of now, I don't really see the need for this PR. It seems to me that advertising UUIDs and Service UUIDs are different things, why should there be a dummy service with the same UUID as the advertising packet ?
> 
> My mindset here was it didn't make sense to advertise a service that didn't actually exist. It seemed a bit nonsensical to me. Which would explain the other two questions as well.

Yeah I guess it mostly comes down to awkward naming in the BLE spec :/ The fact that they refer to everything as services makes things confusing.

> There's another reason, and that's that there are other service disccovery methods that this doesn't take into account (Purely advertising the service). This also includes SDP, where if the service doesn't actually exist, the discovery doesn't happen. This is more of a software problem than a hardware problem, and doesn't _seem_ to be against spec, but may disallow other software from discovering it (Software that is unable to do discovery filtering based on advertised services, and only on discovered services).

I agree, if we ever find a bluetooth stack that has this sort of limitations (only able to filter on service UUIDs instead of advertisement UUIDs) then I'm ok with merging this PR as a hacky workaround but until we do I'd rather keep this out of the code base. :) 

> > Also if there's a canonical source that recommends having a matching service, it seems that a full fix would be to also advertise the other service uuids ?
> 
> In my free time in school, I got a bit curious about this and scoured the BT(LE) specs, and it doesn't mention if either way is needed. In fact, it seems that service advertisement is just an advertisement that happens to have a Service UUID in it, and no other information is provided alongside it (That I can find. I went through GATT, GAP, and a bit of LE specs).

Thanks a lot for taking the time to dive in the spec! That's appreciated :) 

> So I suppose, since it's not mentioned in the spec, this pull request could be ignored, but keep in mind that services that are purely advertised don't show up during actual connection and service discovery. Only actual provided services show up there.

Yep, that's intended behavior, at least in my mind.

Thank you!

**Comment by Noodlez1232** [comment]
Sounds good to me! Yep. I'll go ahead and close it, and we'll re-open it if it comes up. Thanks! It was still a good learning experience.

**Pull request closed without merging**